### PR TITLE
Fix/change reporter claiming mechanism

### PR DESCRIPTION
**Pull request by FoodChain1028**
Please follow the guidelines in [Contribution.md](https://github.com/social-tw/social-tw-website/blob/main/CONTRIBUTING.md) first, then start to create your pull request here.

## Summary

convert `reportNegRep` into `reportNonNullifier` circuits and also changed the related Verifier, VeriferHelper and VerifierHelperManager.

## Details

The circuit is changed from `reportNegRepProof` -> `reportNonNullifierProof`, and delete the `type` part because the `reporter` （檢舉者） actually has the same situation with `poster` (被檢舉者).

i.e.,

* reporter reported in epoch  `R` with epochKey `epkR`
* poster posted in epoch `P` with epochKey `epkP`

While claiming the reputation, they both need to provide their `epoch` and `epk` **AT THAT MOMENT**.

## Impacted Areas

For `ReputationAPI`: (@UranusLin this would help your work :) ) 

檢舉成功：
* Nullifier：
	* 參與檢舉的投票者：`+1`
	```
		claimPosRep(
			publicSignals = reportNullifierProof.publicSignals
			proof = reportNullifierProof.proof
			identifier = reportNullifierVHelper
			change = 1
		)
	```
* Non-Nullifier：
	* 檢舉者：`+3`：
		```
		claimPosRep(
			publicSignals = reportNonNullifierProof.publicSignals
			proof = reportNonNullifierProof.proof
			identifier = reportNonNullifierVHelper
			change = 3
		)
		```
	* 被檢舉者：`-5`:
		```
		claimNegRep(
			publicSignals = reportNonNullifierProof.publicSignals
			proof = reportNonNullifierProof.proof
			identifier = reportNonNullifierVHelper
			change = 5
		)
		```
* 檢舉失敗：
	* Non-Nullifier：
		* 檢舉者：`-1`：
			```
			claimNegRep(
				publicSignals = reportNonNullifierProof.publicSignals
				proof = reportNonNullifierProof.proof
				identifier = reportNonNullifierVHelper
				change = 1
			)
			```

## Tests

see `contracts/test/ClaimPosRep.test.ts` and  `contracts/test/ClaimNegRep.test.ts`

## Verification Steps

`yarn test` (most importantly in circuits and contracts)

## Todo

Break down the progress of the PR for everyone to see what else you intend to include in this PR.

-   [X] Draft implementation for `function`
-   [X] Unit tests for `function`

## Checklist

-   [X] All new and existing tests pass
-   [X] I have updated the documentation (if applicable)
-   [X] My changes do not introduce new security vulnerabilities

### CODE REVIEW

**Comment by FoodChain1028** [comment]
Hi @Xiawpohr could you help me to review this PR?

File: packages/circuits/test/proveReportNonNullifier.test.ts [hunk 1]
```diff
@@ -97,77 +95,51 @@ describe('Prove report negative reputation in Unirep Social-TW', function () {
...
-        expect(controlData.chainId.toString()).to.be.equal(chainId.toString())
-
-        // we don't reveal the nonce, so this is equal to BigInt(0)
-        expect(controlData.nonce.toString()).to.be.equal('0')
+        try {
+            const { isValid } = await genProofAndVerify(circuit, circuitInputs)
+            expect(isValid).to.be.false
+        } catch (error) {
+            console.log('Expected error occurred:\n\n', error)
+        }
```
Comment (Xiawpohr) [inline packages/circuits/test/proveReportNonNullifier.test.ts:113]: Don't recommend to use try-catch in the test cases. If you still want to write try-catch in the test cases, write some assertions in catch block too.

File: packages/circuits/test/proveReportNonNullifier.test.ts [hunk 2]
```diff
@@ -79,12 +78,11 @@ describe('Prove report negative reputation in Unirep Social-TW', function () {
         expect(controlData.nonce.toString()).to.be.equal('0')
     })
 
-    it('should generate a negative reputation proof with type 1 and output with correct epochKey', async () => {
+    it('should revert with invalid userId', async () => {
```
[the remark this answers is not in the document]
Comment (FoodChain1028) [reply packages/circuits/test/proveReportNonNullifier.test.ts:81]: oh I think it should be removed.

File: packages/circuits/src/types/ProofGenerationError.ts
Comment (Xiawpohr) [inline packages/circuits/src/types/ProofGenerationError.ts:1]: I suggest to rename this file to `errors.ts` and put it  in `/src` folder. So in the future you can write all error classes in this file which is close to application codes.

Only type definitions are recommended to put in `/types` folder.

File: packages/circuits/src/types/ProofGenerationError.ts [hunk 3]
```diff
@@ -0,0 +1,9 @@
+export class ProofGenerationError extends Error {
+    constructor(message?: string) {
+        super(message)
+    }
+}
+
+export const InvalidProofInputError = new ProofGenerationError(
```
Comment (Xiawpohr) [inline packages/circuits/src/types/ProofGenerationError.ts:7]: I suggest initializing the error object at the moment the error occurs, and error message should describe the situation at that time. For example,

```js
function foo {
  if (somethingIsWrong) {
    throw new ProofGenerationError("something is wrong")
  }
}
```

**Review by Xiawpohr — APPROVED** [review]
Thanks for your hard works.

**Pull request merged**